Dashboard summary page

fm1psy opened the pull request:
This PR aims to close issue #29

This will allow users to add an email subscribing them to a daily report.

fm1psy committed:
add title and subheader

Commit:
Update scores

fm1psy committed:
add s3 client function

fm1psy committed:
Merge branch 'dashboard-summary-page' of https://github.com/Zhi-704/Railway-Tracker into dashboard-summary-page

Commit:
Update scores

fm1psy committed:
edit client to ses instead of s3

fm1psy committed:
add email submission form

Commit:
Update scores

fm1psy committed:
add s3 client function

Commit:
Update scores

fm1psy committed:
add sql functions
add a db connection function
add a cursor function to execute queries

Commit:
Update scores

fm1psy committed:
add email validity function

fm1psy committed:
Merge branch 'dashboard-summary-page' of https://github.com/Zhi-704/Railway-Tracker into dashboard-summary-page

Commit:
Update scores

fm1psy committed:
add function to check if an email is already in the database

Commit:
Update scores

fm1psy committed:
add subscription functinoality

Commit:
Update scores

fm1psy committed:
clean code base
removed unused functions from the subscriber page

Commit:
Update scores

fm1psy committed:
refactor code
made changes to reach a pylint score of 100
removed faulty tests to make sure github actions would not raise an issue
further testing will be part of a new issue.

Commit:
Update scores

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+"""
+creates a page in the dashboard that allows users to subscribe to a summary report that
+is emailed to them every day. These reports are also stored in an S3 bucket
+"""
+import re
+from os import environ
+from dotenv import load_dotenv
+import streamlit as st
+from psycopg2 import connect
+from psycopg2.extras import RealDictCursor
+from psycopg2.extensions import connection, cursor
```
```suggestion
import re
from os import environ

from dotenv import load_dotenv
import streamlit as st
from psycopg2 import connect
from psycopg2.extras import RealDictCursor
from psycopg2.extensions import connection, cursor
```

A-Faris replied:
```suggestion
"""

import re
from os import environ

from dotenv import load_dotenv
import streamlit as st
from psycopg2 import connect
from psycopg2.extras import RealDictCursor
from psycopg2.extensions import connection, cursor
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+            port=environ['DB_PORT']
+        )
+    except Exception as e: # pylint: disable=broad-exception-caught
+        print(e)
```
```suggestion
        logging.error(e)
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+            password=environ['DB_PASSWORD'],
+            port=environ['DB_PORT']
+        )
+    except Exception as e: # pylint: disable=broad-exception-caught
```
Try to use a more specific error

fm1psy replied:
Unfortunately there are too many possible errors within psycopg2, so it is difficult to account for every error.

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+from psycopg2.extras import RealDictCursor
+from psycopg2.extensions import connection, cursor
+
+def get_db_connection() -> connection:
```
```suggestion
def get_db_connection() -> connection | None:
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+    """return a cursor object based on a given connection"""
+    try:
+        return conn.cursor(cursor_factory=RealDictCursor)
+    except Exception as e: # pylint: disable=broad-exception-caught
```
Use a more specific error

FarihaChoudhury replied:
I understand the difficulties of this

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+    try:
+        return conn.cursor(cursor_factory=RealDictCursor)
+    except Exception as e: # pylint: disable=broad-exception-caught
+        st.write(e)
```
```suggestion
        logging.error(e)
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+
+def is_email_valid(email: str) -> bool:
+    """return a bool based on whether this is a valid email"""
+    email_pattern = re.compile(r"^[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$")
```
```suggestion
    email_pattern = re.compile(r"[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$")
```

A-Faris replied:
https://docs.python.org/3/library/re.html#search-vs-match
since match is used, using ^ at the beginning is redundant unless you want to use search

A-Faris replied:
You can also use fullmatch and remove ^ and $

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+def is_email_valid(email: str) -> bool:
+    """return a bool based on whether this is a valid email"""
+    email_pattern = re.compile(r"^[\w\-\.]+@([\w\-]+\.)+[\w\-]{2,4}$")
+    return re.match(pattern=email_pattern, string=email) is not None
```
```suggestion
    return re.match(email_pattern, email) is not None
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+    return re.match(pattern=email_pattern, string=email) is not None
+
+
+def get_from_subscription_form() -> dict:
+    """create a subscription form to the summary report"""
+    email = ""
+    with st.form("summary_subscription", clear_on_submit=True, border=True):
+        email = st.text_input("Email:")
+        submit = st.form_submit_button()
+
+    if email != "" and submit:
+        if not is_email_valid(email):
+            st.error("This is not a valid email.")
+        else:
+            return {'email': email}
+    return {}
```
```suggestion
def get_from_subscription_form() -> str | None:
    """create a subscription form to the summary report"""
    with st.form("summary_subscription", clear_on_submit=True, border=True):
        email = st.text_input("Email:")
        submit = st.form_submit_button()

    if email and submit:
        if is_email_valid(email):
            return email
        st.error("This is not a valid email.")
    return None
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+            curs.execute("INSERT INTO subscriber (email) VALUES (%s)", (email, ))
+            conn.commit()
+        return 1
+    except Exception as e: # pylint: disable=broad-exception-caught
```
Try to use a more specific error

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+            conn.commit()
+        return 1
+    except Exception as e: # pylint: disable=broad-exception-caught
+        st.write(e)
```
```suggestion
        logging.error(e)
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+    subscriber_input = get_from_subscription_form()
+    conn = get_db_connection()
+    email = subscriber_input.get("email", None)
+    if email and is_email_already_subscribed(email, conn):
+        st.error("This email is already subscribed to the summary report.")
```
```suggestion
    if email and is_email_already_subscribed(email, conn):
        logging.error("This email is already subscribed to the summary report.")
        st.error("This email is already subscribed to the summary report.")
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+    st.title("Summary Report")
+    st.subheader(
+        "Subscribe to a daily report on our tracking results by adding your email below.")
+    subscriber_input = get_from_subscription_form()
+    conn = get_db_connection()
+    email = subscriber_input.get("email", None)
```
```suggestion
    email = get_from_subscription_form()
    conn = get_db_connection()
```

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+    if email and is_email_already_subscribed(email, conn):
+        st.error("This email is already subscribed to the summary report.")
+    elif email:
+        if upload_new_subscriber(conn, email) is not None:
+            st.success(f"{email} is now subscribed to the summary report.")
+        else:
+            st.error(f"An error ocurred when trying to add {email} to the database.")
```
```suggestion
        if upload_new_subscriber(conn, email):
            st.success(f"{email} is now subscribed to the summary report.")
        else:
            st.error(f"An error occurred when trying to add {email} to the database.")
```

A-Faris reviewed (changes requested):
Great foundation, consider logging

A-Faris committed:
Merge branch 'main' into dashboard-summary-page

Commit:
Update scores

A-Faris commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,95 @@
+            st.error(f"An error ocurred when trying to add {email} to the database.")
+
+
+if __name__ == "__main__":
+    load_dotenv()
+    deploy_subscription_page()
```
```suggestion
if __name__ == "__main__":
    load_dotenv()
    logging.basicConfig(level=logging.INFO,
                        format="%(asctime)s - %(levelname)s - %(message)s")
    deploy_subscription_page()
```

fm1psy committed:
changes made for the pull request
edited function return type hints to include None when needed
refactored code so that it would reflect suggestions made
by quality assurance
added logging

Commit:
Update scores

FarihaChoudhury commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,98 @@
+    elif subscriber_email:
+        if upload_new_subscriber(conn, subscriber_email) is not None:
+            st.success(f"{subscriber_email} is now subscribed to the summary report.")
+            logging.info(f"New email added: {subscriber_email}")
```
you can use %s instead of an f string, it will get rid of the pylint issue

FarihaChoudhury commented on dashboard/pages/daily_report.py:
```diff
@@ -0,0 +1,98 @@
+
+
+if __name__ == "__main__":
+    load_dotenv()
```
might need to move this out because when you dockerise it won't be running

fm1psy replied:
I believe that is only the case for lambda functions since they invoke the function rather than run the file as main. Since we're running this as a service I do not think this will cause issues.

FarihaChoudhury replied:
that sounds very true!

fm1psy committed:
further refactoring, missed a few things

Commit:
Update scores

FarihaChoudhury committed:
Merge branch 'main' into dashboard-summary-page

A-Faris reviewed (approved):
Looks good
